### PR TITLE
Incremental Feature Selection Implementation

### PR DESCRIPTION
**Pull request by Nama21yo**
## Description

Implemented `incremental feature selection` with configurable thresholds, redundancy checks, and adaptive binary search for optimal feature subset selection.

## Motivation and Context

Part of the effort to scale the MOSES MVP into supporting multiple deme expansion that depends on feature selection.

## How Has This Been Tested?

* All tests executed locally including edge-cases.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

* [x] My code follows the code style of this project.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.

### CODE REVIEW

File: feature-selection/incremental.metta [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+       (let (mkITable $rows $labels) $itable
+           (let $allIndices (List.range 0 (- (List.length $labels) 1))
+               (chain (calculateMutualInformation $itable $allIndices) $score
+                   ($score (List.listToExpr $allIndices)))))
+       ;; Perform feature selection
+       (chain (if (> $targetSize 0)
+                  (adaptiveFeatureSelection $itable $targetSize $maxInteractionTerms $redundancyThreshold)
+                  (incrementalFeatureSelection $itable $threshold $maxInteractionTerms $redundancyThreshold)) $selectedIndices
+           (chain (if (== (List.length $selectedIndices) 0)
+                      0
```
[the remark this answers is not in the document]
Comment (Yagth) [reply feature-selection/incremental.metta:250]: But I couldn't find this logic in the C++ version. Can you point me to it? Why are we returning `0` if no feature sets are selected?

File: feature-selection/incremental.metta [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+;; For each term size: find relevant features, remove redundant ones, continue with remainder
+;;
+;; Parameters:
+;;   $itable - ITable containing features and target variable
+;;   $threshold - MI threshold for feature relevance (0.0 to 1.0)
+;;   $maxInteractionTerms - Maximum size of feature combinations to consider
+;;   $redundancyThreshold - Multiplier for redundancy detection (0 = no redundancy removal)
+;; Returns:
+;;   List of selected feature indices
+(= (incrementalFeatureSelection $itable $threshold $maxInteractionTerms $redundancyThreshold)
```
Comment (Yagth) [inline feature-selection/incremental.metta:180]: Can you change the name to `incrementalSelection` to make it easier to map with the C++ version?

File: feature-selection/incremental.metta [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+;;   $redundancyThreshold - Intensity of redundancy removal (0=none, 0.5=moderate, 1=aggressive)
+;;
+;; Returns:
+;;   Tuple: (mutual-information-score (selected-feature-indices))
+;;   Example: (0.85 (0 2 4)) means features at indices 0, 2, 4 with MI score 0.85
+;;
+;; Usage Examples:
+;;   !(incrementalFeatureSelection tableXor 0.1 0 2 0.5)    # Threshold-based selection
+;;   !(incrementalFeatureSelection table1 0 5 3 0.4)        # Select exactly 5 features
+(= (incrementalFeatureSelection $itable $threshold $targetSize $maxInteractionTerms $redundancyThreshold)
```
[the remark this answers is not in the document]
Comment (Nama21yo) [reply feature-selection/incremental.metta:238]: gosh I forget adding Types yeah definitely will add that.

File: feature-selection/incremental.metta [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+   (if (and (<= $threshold 0) (<= $targetSize 0))
+       ;; Default behavior: return all features with their combined MI score
+       (let (mkITable $rows $labels) $itable
```
[the remark this answers is not in the document]
Comment (Nama21yo) [reply feature-selection/incremental.metta:241]: That was the way of my writing(should have checked other implementation) I will do it inside function argument .

File: feature-selection/incremental.metta [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+           (let $allIndices (List.range 0 (- (List.length $labels) 1))
+               (chain (calculateMutualInformation $itable $allIndices) $score
+                   ($score (List.listToExpr $allIndices)))))
```
[the remark this answers is not in the document]
Comment (Nama21yo) [reply feature-selection/incremental.metta:244]: Sorry could explain it. Is it the function argument thing or is it something else when you say this should be the case.

File: feature-selection/incremental.metta (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nama21yo) [reply feature-selection/incremental.metta:250]: this is an edge case it hasn't been written explicitly in the c++ but With no features selected, the inputs carry no information about the output. So MI will be 0. That why wrote it in that manner.
